Translation Static-type-checking (#136)
reactjs/es.reactjs.org: content/docs/static-type-checking.md

next: refs-and-the-dom.html
---

Comprobador de tipo estático [Flow](https://flow.org/) y [TypeScript](https://www.typescriptlang.org/) identifican cierto tipo de problemas incluso antes de ejecutar tu código. También pueden mejorar el flujo de trabajo del desarrollador al agregar características como la finalización automática. Por este motivo, recomendamos utilizar Flow o TypeScript en lugar de `PropTypes` para bases de código más grandes.

## Flow {#flow}

[Flow](https://flow.org/) es un comprobador de tipo estático para tu código JavaScript. Desarrollado por Facebook y a menudo usado con React. te permite escribir las variables, funciones y componentes React con una sintaxis de tipo especial, detectando antes los errores. Puedes leer una [Introducción a Flow](https://flow.org/en/docs/getting-started/) para conocer sus conceptos básicos.

Para usar Flow, tu necesitas:

* Agregar Flow como una dependencia a tu proyecto.
* Asegúrate que la sintaxis de Flow se elimina del código compilado.
* Agregue comentarios de tipo y ejecute Flow para verificarlos.

Vamos a explicar estos pasos a continuación en detalle.

### Agregando Flow a tu proyecto {#adding-flow-to-a-project}

> Primero, navegue hasta el directorio donde esta tu proyecto, en la terminal, Deberás ejecutar el siguiente comando:

Reviewer: Primero, navega hasta el directorio donde esta tu proyecto en la terminal. Deberás ejecutar el siguiente comando:

>  Si usas [Yarn](https://yarnpkg.com/), ejecutar:

Reviewer: Si usas [Yarn](https://yarnpkg.com/), ejecuta:

```bash
yarn add --dev flow-bin
```

> Si usas [npm](https://www.npmjs.com/), ejecutar:

Reviewer: Si usas [npm](https://www.npmjs.com/), ejecuta:

```bash
npm run flow init
```

Este comando creará un archivo de configuración de Flow que deberás confirmar (commit).

### Elimminando la sintaxis de Flow del código compilado {#stripping-flow-syntax-from-the-compiled-code}

Flow extiende el lenguaje JavaScript con una sintaxis especial para declaraciones y anotaciones de tipo. Sin embargo, los navegadores no interpretan esta sintaxis, por lo que debes asegurarte de que no termine en el paquete compilado de JavaScript que envías al navegador.

La forma exacta de hacerlo depende de las herramientas que utilices para compilar JavaScript.

#### Comando Create React App {#create-react-app}

Si tu proyecto fue configurado usando el comando [Create React App](https://github.com/facebookincubator/create-react-app), Felicitaciones! Las declaraciones y anotaciones de Flow ya se están eliminando de forma predeterminada, por lo que no necesitas nada más en este paso.

#### Babel {#babel}

>Nota:
>
>Estas instrucciones **NO** son para usuarios que usaron Create React App. aunque la aplicación creada con este comando usa babel, ya esta configurada para entender Flow. Solo siga este paso si **NO** usaste dicho comando.

> Si configuró manualmente Babel para su proyecto, deberás instalar un ajuste preestablecido especial para Flow.

Reviewer 2: Quizá debamos traducir *preset*, quizá no. Por el momento al ser un concepto que debes dominar si usas Babel no debe haber problemas para dejarlo sin traducir. Además aún no hemos discutido el término más apropiado e incluso tú has usado 2 traducciones distintas en la misma página.
```suggestion
Si configuraste manualmente Babel en tu proyecto, deberás instalar un *preset* especial para Flow.
```

Si usas Yarn, ejecuta:

```bash
yarn add --dev babel-preset-flow
```

Si usas npm, ejecuta:

```bash
npm install --save-dev babel-preset-flow
```

> Luego agrega el preajuste `flow` a tu [Configuración de Babel](https://babeljs.io/docs/usage/babelrc/). Por ejemplo, si configuraste Babel a través del archivo `.babelrc`, podría verse así:

Reviewer 2: ```suggestion
Luego agrega el *preset* `flow` a tu [configuración de Babel](https://babeljs.io/docs/usage/babelrc/). Por ejemplo, si configuraste Babel a través del archivo `.babelrc`, podría verse así:
```

}
```

Esto te permitirá usar la sintaxis de Flow en tu código.

>Nota:
>

> >Flow no requiere el preajuste `react`, pero a menudo se usan juntos. Flow en si mismo entiende la sintaxis JSX fuera de la caja.

Reviewer 2: Esto no suena muy natural en español.
```suggestion
>Flow no requiere el *preset* `react`, pero a menudo se usan juntos. Flow es capaz de entender la sintaxis JSX sin configuración adicional.
```

#### Otras configuraciones {#other-build-setups}

> Si no usaste el comando Create React App o Babel, puedes usar [flow-remove-types](https://github.com/flowtype/flow-remove-types) para eliminar las anotaciones de tipo.

Reviewer 2: ```suggestion
Si no usaste Create React App o Babel, puedes usar [flow-remove-types](https://github.com/flowtype/flow-remove-types) para eliminar las anotaciones de tipos.
```

### Corriendo Flow {#running-flow}

Si seguiste las instrucciones anteriores, deberías poder ejecutar Flow por primera vez.

Si usas Yarn, ejecuta:

```bash
yarn flow
```

Si usas npm, ejecuta:

```bash
npm run flow
```

Deberías ver un mensaje como el siguiente:

```
No errors!
✨  Done in 0.17s.
```

### Agregando anotaciones de tipo Flow {#adding-flow-type-annotations}

Por defecto, Flow solo verifica los archivos que incluyen esta anotación:

```js
// @flow
```

> Nomalmente se coloca en la parte superior de un archivo. Intente agregarlo a algunos archivos en tu proyecto y ejecute `yarn flow` o `npm run flow` para que mires si Flow ya encontró algún problema.

Reviewer 2: ```suggestion
Nomalmente se coloca en la parte superior de un archivo. Intenta agregarlo a algunos archivos en tu proyecto y ejecuta `yarn flow` o `npm run flow` para que mires si Flow ya encontró algún problema.
```

> También hay [una opción](https://flow.org/en/docs/config/options/#toc-all-boolean) para forzar a Flow a verificar **todos** los archivos independientemente de la anotación. Esto puede ser demaciado ruidoso para los proyectos existentes, pero es razonable para un nuevo proyecto si desea escribirlo completamente con Flow.

Reviewer 2: ```suggestion
También hay [una opción](https://flow.org/en/docs/config/options/#toc-all-boolean) para forzar a Flow a verificar **todos** los archivos independientemente de la anotación. Esto puede ser demasiado ruidoso para los proyectos existentes, pero es razonable para un nuevo proyecto si deseas escribirlo completamente con Flow.
```

Ahora estás listo! Te recomendamos consultar los siguientes recursos para obtener más información sobre Flow:

> * [Documentación Flow: Anotaciones de tipo](https://flow.org/en/docs/types/)

Reviewer 2: ```suggestion
* [Documentación de Flow: Anotaciones de tipos](https://flow.org/en/docs/types/)
```

> * [Documentación Flow: Editores](https://flow.org/en/docs/editors/)

Reviewer 2: ```suggestion
* [Documentación de Flow: Editores](https://flow.org/en/docs/editors/)
```

> * [Documentación Flow: React](https://flow.org/en/docs/react/)

Reviewer 2: ```suggestion
* [Documentación de Flow: React](https://flow.org/en/docs/react/)
```

* [Linting in Flow](https://medium.com/flow-type/linting-in-flow-7709d7a7e969)

## TypeScript {#typescript}

> [TypeScript](https://www.typescriptlang.org/) Es un lenguaje de programación desarrollado por Microsoft.  Es un superconjunto escrito de JavaScript e incluye su propio compilador. Al ser un lenguaje escrito, TypeScript puede detectar errores y fallos en el momento de la creación, mucho antes de que tu aplicación entre en funcionamiento. Puedes obtener más información sobre el uso de TypeScript con React [Aquí](https://github.com/Microsoft/TypeScript-React-Starter#typescript-react-starter).

Reviewer 2: ```suggestion
[TypeScript](https://www.typescriptlang.org/) Es un lenguaje de programación desarrollado por Microsoft.  Es un superconjunto JavaScript con tipos e incluye su propio compilador. Al ser un lenguaje con tipos, TypeScript puede detectar errores y fallos en el momento de la creación, mucho antes de que tu aplicación entre en funcionamiento. Puedes obtener más información sobre el uso de TypeScript con React [aquí](https://github.com/Microsoft/TypeScript-React-Starter#typescript-react-starter).
```

Para usar TypeScript, necesitas:
* Agregar TypeScript como una dependencia en tu proyecto.
* Configurar las opciones del compilador de TypeScript.

> * Usa las extensiones de archivo correctas.

Reviewer 2: ```suggestion
* Usar las extensiones de archivo correctas.
```

> * Agregue definiciones para las bibliotecas que usa.

Reviewer 2: ```suggestion
* Agregar definiciones para las bibliotecas que usas.
```

Repasemos esto en detalle..

### Usando TypeScript con Create React App {#using-typescript-with-create-react-app}

> El comando Create React App soporta TypeScript fuera de la caja.

Reviewer 2: https://www.fundeu.es/recomendacion/soporte-soportar/
```suggestion
Create React App es compatible con TypeScript sin necesidad de configuración adicional.
```

> Para crear un **nuevo proyecto** con soporte TypeScript, ejecute:

Reviewer 2: ```suggestion
Para crear un **nuevo proyecto** compatible con TypeScript, ejecuta:
```

```bash
npx create-react-app my-app --typescript
```

> También puedes agregarlo a un **proyecto existente Create React App**, [documentación aquí](https://facebook.github.io/create-react-app/docs/adding-typescript).

Reviewer 2: ```suggestion
También puedes agregarlo a un **proyecto de Create React App existente**, [documentación aquí](https://facebook.github.io/create-react-app/docs/adding-typescript).
```

>Nota:
>

> >Si tu usas Create React App, tu puedes **omitir el resto de esta página**. Describe la configuración manual que no se aplica a los usuarios de este comando.

Reviewer 2: ```suggestion
>Si usas Create React App, puedes **omitir el resto de esta página**. Describe la configuración manual que no se aplica a los usuarios de este comando.
```

### Agregando TypeScript a tu proyecto {#adding-typescript-to-a-project}
Todo comienza con ejecutar un comando en tu terminal.

> Si tu usas [Yarn](https://yarnpkg.com/), ejecuta:

Reviewer 2: ```suggestion
Si usas [Yarn](https://yarnpkg.com/), ejecuta:
```

```bash
yarn add --dev typescript
```

> Si tu usas [npm](https://www.npmjs.com/), ejecuta:

Reviewer 2: ```suggestion
Si usas [npm](https://www.npmjs.com/), ejecuta:
```

```bash
npm install --save-dev typescript
```

> Felicidades Has instalado la última versión de TypeScript en tu proyecto. Instalar TypeScript nos da acceso al comando `tsc`. Antes de la configuración, agreguemos `tsc` a la sección de" scripts "en nuestro `package.json`:

Reviewer 2: ```suggestion
¡Felicidades! Has instalado la última versión de TypeScript en tu proyecto. Instalar TypeScript nos da acceso al comando `tsc`. Antes de la configuración, agreguemos `tsc` a la sección de "scripts" en nuestro `package.json`:
```

}
```

### Configurando el compilador de TypeScript {#configuring-the-typescript-compiler}

> El compilador no nos sirve de nada hasta que le decimos qué hacer. En TypeScript, estas reglas se definen en un archivo especial llamado `tsconfig.json`. Para generar este archivo ejecute:

Reviewer 2: ```suggestion
El compilador no nos sirve de nada hasta que le decimos qué hacer. En TypeScript, estas reglas se definen en un archivo especial llamado `tsconfig.json`. Para generar este archivo ejecuta:
```

```bash
tsc --init
```

> Mirando el `tsconfig.json` generado ahora, puede ver que hay muchas opciones que puede usar para configurar el compilador. Para obtener una descripción detallada de todas las opciones, consulte [aquí](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html).

Reviewer 2: ```suggestion
Mirando el `tsconfig.json` generado ahora, puedes ver que hay muchas opciones que puedes usar para configurar el compilador. Para obtener una descripción detallada de todas las opciones, consulta [aquí](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html).
```

}
```

¡Genial! Ahora, cuando ejecutamos nuestro script de compilación, el compilador enviará el javascript generado a la carpeta `build`. El [TypeScript React Starter](https://github.com/Microsoft/TypeScript-React-Starter/blob/master/tsconfig.json) proporciona un `tsconfig.json` con un buen conjunto de reglas para comenzar.

En general, no deseas mantener el javascript generado en tu control de código fuente, así que asegúrate de agregar la carpeta de compilación a tu `.gitignore`.

### Extensiones de archivo {#file-extensions}
En React, lo más probable es que escribas tus componentes en un archivo `.js`. En TypeScript tenemos 2 extensiones de archivo.:

`.ts` es la extensión de archivo predeterminada, mientras que `.tsx` es una extensión especial que se usa para los archivos que contienen `JSX`.

### Corriendo TypeScript {#running-typescript}

Si seguiste las instrucciones anteriores, deberías poder ejecutar TypeScript por primera vez.

```bash
yarn build
```

Si usas npm, ejecuta:

```bash
npm run build
```

Si no ves ninguna salida, significa que se completó correctamente.

> ### Tipo de definiciones {#type-definitions}

Reviewer 2: ```suggestion
### Definiciones de tipos {#type-definitions}
```

> Para poder mostrar errores y sugerencias de otros paquetes, el compilador se basa en archivos de declaración. Un archivo de declaración proporciona toda la información de tipo sobre una biblioteca. Esto nos permite usar bibliotecas javascript como las de npm en nuestro proyecto.

Reviewer 2: ```suggestion
Para poder mostrar errores y sugerencias de otros paquetes, el compilador se basa en archivos de declaración. Un archivo de declaración proporciona toda la información de tipos sobre una biblioteca. Esto nos permite usar bibliotecas javascript como las que están en npm en nuestro proyecto.
```

Hay dos formas principales de obtener declaraciones para una biblioteca.:

> __Bundled__ - La biblioteca tiene sus propios archivos de declaración bundles. Esto es genial para nosotros, ya que todo lo que tenemos que hacer es instalar la biblioteca, y podemos usarla de inmediato. Para verificar si una biblioteca tiene tipos bundles, busque un archivo `index.d.ts` en el proyecto. Algunas bibliotecas lo tendrán especificado en su `package.json` bajo el campo `typings` o `types`.

Reviewer 2: ```suggestion
__Integradas__ - La biblioteca incluye sus propios archivos de declaración. Esto es genial para nosotros, ya que todo lo que tenemos que hacer es instalar la biblioteca y podemos usarla de inmediato. Para verificar si una biblioteca tiene tipos integrados, busca un archivo `index.d.ts` en el proyecto. Algunas bibliotecas lo tendrán especificado en su `package.json` bajo el campo `typings` o `types`.
```

> __[DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)__ - DefinitelyTyped es un gran repositorio de declaraciones para bibliotecas que no incluyen un archivo bundle. Las declaraciones son de carácter público y son administradas por Microsoft y colaboradores de código abierto. React, por ejemplo, no incluye su propio archivo de declaración. En su lugar, podemos obtenerlo de DefinitelyTyped. Para ello ingresa este comando en tu terminal.

Reviewer 2: ```suggestion
__[DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped)__ - DefinitelyTyped es un gran repositorio de declaraciones para bibliotecas que no incluyen un archivo de declaraciones. Las declaraciones son de carácter público y son administradas por Microsoft y colaboradores de código abierto. React, por ejemplo, no incluye su propio archivo de declaración. En su lugar, podemos obtenerlo de DefinitelyTyped. Para ello ingresa este comando en tu terminal.
```

```

> __Local Declarations__

Reviewer 2: ```suggestion
__Declaraciones locales__
```

> A veces, el paquete que desea utilizar no incluye declaraciones bundle ni está disponible en DefinitelyTyped. En ese caso, podemos tener un archivo de declaración local. Para hacer esto, cree un archivo `declarations.d.ts` en la raíz de tu directorio de origen. Una simple declaración podría verse así:

Reviewer 2: ```suggestion
A veces, el paquete que deseas utilizar no incluye declaraciones ni está disponible en DefinitelyTyped. En ese caso, podemos tener un archivo de declaración local. Para hacer esto, crea un archivo `declarations.d.ts` en la raíz de tu directorio de origen. Una simple declaración podría verse así:
```

}
```

¡Ya estás listo para codificar! Recomendamos consultar los siguientes recursos para obtener más información sobre TypeScript:

> * [Documentación TypeScript: Tipos Basicos](https://www.typescriptlang.org/docs/handbook/basic-types.html)

Reviewer 2: ```suggestion
* [Documentación de TypeScript: Tipos Básicos](https://www.typescriptlang.org/docs/handbook/basic-types.html)
```

> * [Documentación TypeScript: Migración desde Javascript](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html)

Reviewer 2: ```suggestion
* [Documentación de TypeScript: Migración desde Javascript](https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html)
```

> * [Documentación TypeScript: React y Webpack](https://www.typescriptlang.org/docs/handbook/react-&-webpack.html)

Reviewer 2: ```suggestion
* [Documentación de TypeScript: React y Webpack](https://www.typescriptlang.org/docs/handbook/react-&-webpack.html)
```

## Reason {#reason}

[Reason](https://reasonml.github.io/) no es un lenguaje nuevo; Es una nueva sintaxis y una cadena de herramientas desarrollada por el lenguaje probado en batalla, [OCaml](https://ocaml.org/). Reason le da a OCaml una sintaxis familiar orientada a los programadores de JavaScript, y se dirige a los flujos de trabajo existentes de NPM/Yarn que ya conocen.

> Reason se desarrolla en Facebook y se usa en algunos de sus productos como Messenger. Todavía es un tanto experimental, pero tiene [enlaces de React dedicados](https://reasonml.github.io/reason-react/) mantenidos por Facebook y una [comunidad vibrante](https://reasonml.github.io/docsen/community.html).

Reviewer 2: Este uso de binding no tiene una traducción ampliamente aceptada, quizá conectar pueda funcionar aquí.
https://es.wikipedia.org/wiki/Binding
https://spanish.stackexchange.com/questions/15534/binding-en-castellano
```suggestion
Reason se desarrolla en Facebook y se usa en algunos de sus productos como Messenger. Todavía es un tanto experimental, pero tiene [*bindings* para React](https://reasonml.github.io/reason-react/) mantenidos por Facebook y una [comunidad vibrante](https://reasonml.github.io/docsen/community.html).
```

## Kotlin {#kotlin}

[Kotlin](https://kotlinlang.org/) es un lenguaje de tipo estático desarrollado por JetBrains. Sus plataformas de destino incluyen JVM, Android, LLVM y [JavaScript](https://kotlinlang.org/docs/reference/js-overview.html).

> JetBrains desarrolla y mantiene varias herramientas específicamente para la comunidad React: [React bindings](https://github.com/JetBrains/kotlin-wrappers) así como [Create React Kotlin App](https://github.com/JetBrains/create-react-kotlin-app). Este último te ayuda a comenzar a crear aplicaciones React con Kotlin sin configuración de compilación.

Reviewer 2: ```suggestion
JetBrains desarrolla y mantiene varias herramientas específicamente para la comunidad de React: [*bindings* para React](https://github.com/JetBrains/kotlin-wrappers), así como [Create React Kotlin App](https://github.com/JetBrains/create-react-kotlin-app). Este último te ayuda a comenzar a crear aplicaciones React con Kotlin sin configuración de compilación.
```

## Otros lenguajes {#other-languages}

> Tenga en cuenta que hay otros lenguajes de tipo estático que se compilan en JavaScript y, por lo tanto, son compatibles con React. Por ejemplo, [F#/Fable](http://fable.io)with [elmish-react](https://elmish.github.io/react). Visite sus respectivos sitios para obtener más información y siéntete libre de agregar más idiomas tipificados estáticamente que funcionan con React a esta página!

Reviewer 2: Tipado es un neologismo, pero es ampliemente usado en la literatura. Creo que podríamos usarlo aquí.
```suggestion
Ten en cuenta que hay otros lenguajes de tipo estático que se compilan en JavaScript y por lo tanto, son compatibles con React. Por ejemplo, [F#/Fable](http://fable.io)with [elmish-react](https://elmish.github.io/react). Visita sus respectivos sitios para obtener más información y siéntete libre de agregar más lenguajes estáticamente tipados que funcionan con React a esta página!
```
